hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala

```diff
@@ -572,34 +591,54 @@ class JSBuilder(using TL, State, Ctx) extends CodeBuilder:
         then "./" + os.Path(path).relativeTo(wd).toString
         else path
       doc"""import ${getVar(i._1, N)} from "${relPath}";"""
-    imps.mkDocument(doc" # ") :/: block(p.main, endSemi = false).stripBreaks :: (
+    imps.mkDocument(doc" # ") :/:
+    nonNestedScoped(p.main)(block(_, endSemi = false)).stripBreaks ::
+    locally:
       exprt match
-        case S(sym) => doc"\nlet ${sym.nme} = ${scope.lookup_!(sym, sym.toLoc)}; export default ${sym.nme};\n"
-        case N => doc""
-      )
+      case S(sym) => doc"\nlet ${sym.nme} = ${scope.lookup_!(sym, sym.toLoc)}; export default ${sym.nme};\n"
+      case N => doc""
   
   def worksheet(p: Program)(using Raise, Scope): (Document, Document) =
     reserveNames(p)
     lazy val imps = p.imports.map: i =>
       doc"""${getVar(i._1, N)} = await import("${i._2.toString}").then(m => m.default ?? m);"""
-    blockPreamble(p.imports.map(_._1).toSeq ++ p.main.definedVars.toSeq) ->
+    p.main match
+    case Scoped(syms, body) =>
+      blockPreamble(p.imports.map(_._1).toSeq ++ syms.toSeq) ->
+      (imps.mkDocument(doc" # ") :/: block(body, endSemi = false).stripBreaks)
+    case _ =>
+      blockPreamble(p.imports.map(_._1).toSeq ++ p.main.definedVarsNoScoped.toSeq) ->
       (imps.mkDocument(doc" # ") :/: returningTerm(p.main, endSemi = false).stripBreaks)
-  
-  def blockPreamble(ss: Iterable[Symbol])(using Raise, Scope): Document =
-    // TODO document: mutable var assnts require the lookup
-    val vars = ss.filter(scope.lookup(_).isEmpty).toArray.sortBy(_.uid).iterator.map(l =>
-      l -> scope.allocateName(l))
+
+  def genLetDecls(vars: Iterator[(Symbol, Str)], isScoped: Bool): Document =
     if vars.isEmpty then doc"" else
       doc" # let " :: vars.map: (_, nme) =>
         nme
       .toList.mkDocument(", ")
       :: doc";"
   
+  def blockPreamble(ss: Iterable[Symbol])(using Raise, Scope): Document =
+    // TODO document: mutable var assnts require the lookup
+    val vars = ss.filter(scope.lookup(_).isEmpty).toArray.sortBy(_.uid).iterator.map(l =>
```

Comment:
With this PR, the ugly `lookup` should no longer be needed. I saw that removing it currently crashes the compilation tests. Any idea why?

Reply:
I think this is mainly because in `blockk` function, we have `val pre = blockPreamble(t.definedVarsNoScoped)`. Then `t.definedVarsNoScoped` will also cover symbols that have been collected by the outer `Scoped` object. Without the `filtering` will fail the assertion when we try to duplicate those symbols.